torch/distributed/_spmd/graph_optimization.py

```diff
@@ -481,3 +482,44 @@ def schedule_comm_wait(gm: IterGraphModule) -> None:
                 break
         assert wait_idx >= 0
         gm.graph.move_before(allreduce.node_list[wait_idx:], target_node)
+
+
+@graph_optimization_pass(run_after=tuple())
+def remove_copy_from_optimizer(gm: IterGraphModule) -> None:
+    MAX_COPY_DISTANCE = 5
+    remove_candidates: Set[fx.Node] = set()
+    for node in reversed(gm.graph.nodes):
+        if node.users:
+            continue
```

Comment:
> @mrshenli's internal review: curious, this is the same as DCE? but specialized for foreach? If so, will it work if we just do DCE before going through sparse expansion? cc., @yifuwang 

Not the same as DCE. `fx.Graph` marks `copy_` as side effect node. So a `copy_` won't be removed even if it has no users. That's why we are seeing the orphan gradients `copy_`. However, this pass exists for if and only if the graph has other `copy_` which has zero users and we want to preserve them. Otherwise, we can simply the pass by simply traverse all the `copy_` nodes and remove those which have zero users. And I also think it makes sense to do this after the expansion. I only need this pass to continue the optimizer work at this moment.

Reply:
> fx.Graph marks copy_ as side effect node. 

Oh I see, this is new to me